File: Day244_2025/Q244_Leetcode.java
import java.util.*;

class Q244_Leetcode {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        // Lambda to calculate the profit of adding an extra student
        PriorityQueue<double[]> pq = new PriorityQueue<>(
            (a, b) -> Double.compare(b[0], a[0])
        );

        double total = 0.0;
        for (int[] c : classes) {
            double pass = c[0], totalCls = c[1];
            double profit = (pass + 1) / (totalCls + 1) - pass / totalCls;
            pq.offer(new double[]{profit, pass, totalCls});
            total += pass / totalCls;
        }

        while (extraStudents-- > 0) {
            double[] top = pq.poll();
            total += top[0]; // Add the profit
            double pass = top[1] + 1;
            double totalCls = top[2] + 1;
            double profit = (pass + 1) / (totalCls + 1) - pass / totalCls;
            pq.offer(new double[]{profit, pass, totalCls});
        }

        return total / classes.length;
    }
}

File: Day244_2025/Q244_Leetcode.py
import heapq

class Solution:
    def maxAverageRatio(self, classes: list[list[int]], extraStudents: int) -> float:
        # Max-heap using negative profit
        def profit(pass_cnt, total_cnt):
            return (pass_cnt + 1) / (total_cnt + 1) - pass_cnt / total_cnt

        heap = []
        total = 0.0

        for p, t in classes:
            total += p / t
            heapq.heappush(heap, (-profit(p, t), p, t))  # negative for max-heap

        for _ in range(extraStudents):
            pfit, p, t = heapq.heappop(heap)
            total += -pfit
            p += 1
            t += 1
            heapq.heappush(heap, (-profit(p, t), p, t))

        return total / len(classes)
